1_Download_seqs.R

```diff
@@ -5,86 +5,76 @@
-                                    c(i, 
-                                      entrez_summary(db = "taxonomy", 
-                                                     id = i)$scientificname))}
-      }
-      fasta <- str_split(all_recs_list, pattern = "\n")  # this step is to store the output for easier extraction 
-      if(isEmpty(fasta) == FALSE) {
-        names_v <- vector()
-        fasta_list <- list()
-        count <- 1
-        for(i in 1:length(fasta)){
-          fasta_temp <- fasta[[i]]
-          idx <- str_detect(fasta_temp, ">")
-          if(sum(idx) == 1){
-            names_v[count] <- fasta_temp[idx]
-            fasta_list[[count]] <- str_c(fasta_temp[!idx], collapse = "")
-            count <- count + 1
-          }
-          if(sum(idx) > 1){
-            idx2 <- which(idx)
-            idx2 <- c(idx2, length(idx) + 1)
-            j <- 2
-            for(j in 1:(length(idx2) - 1)){
-              names_v[count] <- fasta_temp[idx2[j]]
-              fasta_list[[count]] <- str_c(fasta_temp[(idx2[j] + 1):(idx2[j+1] - 1)], collapse = "")
-              count <- count + 1
-            }
-          }
-        }
-        names(fasta_list) <- str_sub(names_v, start = 2)
+  ID_n <- length(r_search$ids) # reducing repetitions by reusing this variable
+  cat("\nNumber of ", search_term, " IDs in NCBI taxonomy database: ", ID_n, "\n")
+  if(ID_n != 0) {
+    # linking all taxid with nuccore data 
+    link_history <- entrez_link(dbfrom = "taxonomy", 
+                                db = "nuccore",
+                                web_history = r_search$web_history,
+                                cmd = "neighbor_history")
+    # get the info of all linked noccure records
+    summary_result <- entrez_summary("nuccore", 
+                                     web_history = link_history$web_histories$taxonomy_nuccore, retmode = "xml")
+    meta <- lapply(summary_result, function(x) x[1:27]) # it seems that the first 27 element always the same info
+    #changing NULL to NA https://stackoverflow.com/questions/22870198/is-there-a-more-efficient-way-to-replace-null-with-na-in-a-list
+    nullToNA <- function(x) {
+      x[sapply(x, is.null)] <- NA
+      return(x)
+    }
+    meta2 <- lapply(meta, nullToNA)
+    meta3 <- lapply(meta2, unlist) 
+    meta_df <- as.data.frame(do.call(rbind, meta3))
+    # download sequences in batches using web history data
+    max_seq <- nrow(meta_df)
+    seq_start <- seq(1,max_seq,50)
+    batch_n <- length(seq_start)
+    for(j in 1:batch_n){
+      recs <- entrez_fetch(db="nuccore", web_history=link_history$web_histories$taxonomy_nuccore,
+                           rettype="fasta", retmax=50, retstart=seq_start[j]-1)
+      cat(recs, file=paste0(path_to_output_dir, search_term, ".fasta"), append=TRUE)
+      if(j < batch_n){
+        cat("\n",seq_start[j]+49, "sequences downloaded\r")
+      }else{
+        cat("\n", max_seq, "sequences downloaded\r")
```

Comment:
I found that the API warning message cuts off this final message printing the total number of sequences downloaded - I think this can simply be fixed by changing `sequences downloaded\r` to `sequences downloaded\n`